File: python/zpy/src/zingg_py/bayesian_logreg.py
"""Bayesian logistic regression — port of ``BayesianLogReg.scala``.

Gaussian prior, fit by Newton/IRLS to the MAP, with a Laplace (Gaussian)
approximation of the posterior around it. Keeps the full posterior covariance
``Sigma = H^-1`` so each prediction carries an epistemic variance; accepts a
non-zero prior mean; well-defined with one class or zero labels (the prior
regularizes). Breeze -> numpy, otherwise identical.
"""

from __future__ import annotations

import numpy as np

_LOG2 = np.log(2.0)


def _sigmoid(x):
    return 1.0 / (1.0 + np.exp(-x))


class Fitted:
    """A fitted Laplace posterior ``N(w, sigma)`` over the augmented weights."""

    def __init__(self, w: np.ndarray, sigma: np.ndarray):
        self.w = w
        self.sigma = sigma

    def latent(self, psi_star: np.ndarray) -> tuple[float, float]:
        """Latent score ``(mean, variance)`` for an augmented design vector."""
        mu = float(self.w @ psi_star)
        v = float(psi_star @ (self.sigma @ psi_star))
        return mu, max(v, 0.0)

    def predict_prob(self, psi_star: np.ndarray) -> float:
        """Posterior predictive match probability (MacKay probit approx)."""
        mu, v = self.latent(psi_star)
        return float(_sigmoid(mu / np.sqrt(1.0 + np.pi * v / 8.0)))

    def bald(self, psi_star: np.ndarray) -> float:
        """BALD acquisition: mutual information between label and weights."""
        mu, v = self.latent(psi_star)
        p = _sigmoid(mu / np.sqrt(1.0 + np.pi * v / 8.0))
        marginal = _binary_entropy(p)
        c2 = np.pi * _LOG2 / 2.0
        conditional = np.sqrt(c2 / (v + c2)) * np.exp(-(mu * mu) / (2.0 * (v + c2)))
        return float(marginal - conditional)


def _binary_entropy(p: float) -> float:
    eps = 1e-12
    q = min(1.0 - eps, max(eps, p))
    return -(q * np.log2(q) + (1.0 - q) * np.log2(1.0 - q))


class BayesianLogReg:
    def __init__(
        self,
        prior_mean: np.ndarray,
        prior_prec: np.ndarray,
        max_iter: int = 50,
        tol: float = 1e-8,
        jitter: float = 1e-10,
    ):
        prior_mean = np.asarray(prior_mean, dtype=float)
        prior_prec = np.asarray(prior_prec, dtype=float)
        assert prior_mean.shape[0] == prior_prec.shape[0], (
            f"prior mean dim {prior_mean.shape[0]} != precision dim {prior_prec.shape[0]}"
        )
        self.prior_mean = prior_mean
        self.prior_prec = prior_prec
        self.max_iter = max_iter
        self.tol = tol
        self.jitter = jitter
        self.d = prior_mean.shape[0]

    def fit(self, psi: np.ndarray, y: np.ndarray) -> Fitted:
        """Fit the MAP and return the Laplace posterior.

        ``psi`` is ``n x d``; ``y`` the 0/1 labels. ``n == 0`` returns the prior.
        """
        psi = np.asarray(psi, dtype=float).reshape(-1, self.d)
        y = np.asarray(y, dtype=float).reshape(-1)
        assert psi.shape[1] == self.d, f"feature dim {psi.shape[1]} != prior dim {self.d}"
        n = psi.shape[0]
        w = self.prior_mean.copy()
        for _ in range(self.max_iter):
            g, h = self._grad_hess(psi, y, w, n)
            step = np.linalg.solve(h, g)
            w = w - step
            if np.linalg.norm(step) < self.tol:
                break
        _, h_final = self._grad_hess(psi, y, w, n)
        return Fitted(w, np.linalg.inv(h_final))

    def _grad_hess(self, psi, y, w, n):
        prior_grad = self.prior_prec * (w - self.prior_mean)
        base_h = np.diag(self.prior_prec) + np.eye(self.d) * self.jitter
        if n == 0:
            return prior_grad, base_h
        eta = psi @ w
        s = _sigmoid(eta)
        rw = s * (1.0 - s)
        grad = psi.T @ (s - y) + prior_grad
        h = base_h + (psi.T * rw) @ psi
        return grad, h

File: zscala/src/main/scala/zingg/zs/BayesianLogReg.scala
package zingg.zs

import breeze.linalg.{DenseMatrix, DenseVector, diag, inv, norm}
import breeze.numerics.sigmoid

/** Bayesian logistic regression with a Gaussian prior, fit by Newton/IRLS to
  * the MAP and a Laplace (Gaussian) approximation of the posterior around it.
  *
  * Unlike Spark's `LogisticRegression` this keeps the full posterior covariance
  * `Σ = H⁻¹` (so each prediction carries an epistemic variance), accepts a
  * non-zero prior mean (so the unsupervised similarity proxy can act as the
  * prior), and is well-defined with one class or even zero labels (the prior
  * regularizes). That is what lets a single model drive both cold start and
  * steady-state pair selection — see [[InteractiveSession]].
  *
  * @param priorMean `m₀`, the prior mean over the augmented weights
  * @param priorPrec the diagonal of the prior precision `A` (= `Σ₀⁻¹`)
  */
final class BayesianLogReg(
    priorMean: DenseVector[Double],
    priorPrec: DenseVector[Double],
    maxIter: Int = 50,
    tol: Double = 1e-8,
    jitter: Double = 1e-10
) {
  require(priorMean.length == priorPrec.length,
    s"prior mean dim ${priorMean.length} != precision dim ${priorPrec.length}")

  private val d = priorMean.length

  /** Fit the MAP and return the Laplace posterior. `psi` is `n x d` (each row an
    * augmented design vector), `y` the `0/1` labels. `n == 0` returns the prior
    * unchanged. */
  def fit(psi: DenseMatrix[Double], y: DenseVector[Double]): BayesianLogReg.Fitted = {
    require(psi.cols == d, s"feature dim ${psi.cols} != prior dim $d")
    val n = psi.rows
    var w = priorMean.copy
    var iter = 0
    var done = false
    while (iter < maxIter && !done) {
      val (g, h) = gradHess(psi, y, w, n)
      val step = h \ g
      w = w - step
      done = norm(step) < tol
      iter += 1
    }
    val (_, hFinal) = gradHess(psi, y, w, n)
    new BayesianLogReg.Fitted(w, inv(hFinal))
  }

  private def gradHess(
      psi: DenseMatrix[Double],
      y: DenseVector[Double],
      w: DenseVector[Double],
      n: Int
  ): (DenseVector[Double], DenseMatrix[Double]) = {
    val priorGrad = priorPrec *:* (w - priorMean)
    val baseH     = diag(priorPrec) + (DenseMatrix.eye[Double](d) * jitter)
    if (n == 0) (priorGrad, baseH)
    else {
      val eta = psi * w
      val s   = sigmoid(eta)
      val rw  = s *:* (DenseVector.ones[Double](n) - s)
      val grad = (psi.t * (s - y)) + priorGrad
      val h = baseH.copy
      var r = 0
      while (r < n) {
        val row = psi(r, ::).t
        h += (row * row.t) * rw(r)
        r += 1
      }
      (grad, h)
    }
  }
}

object BayesianLogReg {

  /** A fitted Laplace posterior `N(w, sigma)` over the augmented weights. */
  final class Fitted(val w: DenseVector[Double], val sigma: DenseMatrix[Double]) {

    /** Latent score `(mean, variance)` for an augmented design vector. */
    def latent(psiStar: DenseVector[Double]): (Double, Double) = {
      val mu = w dot psiStar
      val v  = psiStar dot (sigma * psiStar)
      (mu, math.max(v, 0.0))
    }

    /** Posterior predictive match probability (MacKay probit approximation). */
    def predictProb(psiStar: DenseVector[Double]): Double = {
      val (mu, v) = latent(psiStar)
      sigmoid(mu / math.sqrt(1.0 + math.Pi * v / 8.0))
    }

    /** BALD acquisition: mutual information between the label and the weights
      * (Houlsby et al. closed form). High when the pair is near the decision
      * boundary *and* the model is epistemically uncertain there; → 0 once the
      * region is well-determined, regardless of how ambiguous the label is. */
    def bald(psiStar: DenseVector[Double]): Double = {
      val (mu, v) = latent(psiStar)
      val p = sigmoid(mu / math.sqrt(1.0 + math.Pi * v / 8.0))
      val marginal = binaryEntropy(p)
      val c2 = math.Pi * math.log(2.0) / 2.0
      val conditional = math.sqrt(c2 / (v + c2)) * math.exp(-(mu * mu) / (2.0 * (v + c2)))
      marginal - conditional
    }
  }

  private def binaryEntropy(p: Double): Double = {
    val eps = 1e-12
    val q = math.min(1.0 - eps, math.max(eps, p))
    -(q * log2(q) + (1.0 - q) * log2(1.0 - q))
  }

  private def log2(x: Double): Double = math.log(x) / math.log(2.0)
}
